src/operator/first.ts

```diff
@@ -60,9 +60,14 @@ export function first<T, R>(predicate?: (value: T, index: number, source: Observ
 
 export interface FirstSignature<T> {
   (predicate?: (value: T, index: number, source: Observable<T>) => boolean): Observable<T>;
+  <S extends T>(predicate?: (value: T, index: number, source: Observable<T>) => value is S): Observable<S>;
   (predicate: (value: T, index: number, source: Observable<T>) => boolean, resultSelector: void, defaultValue?: T): Observable<T>;
+  <S extends T>(predicate: (value: T, index: number, source: Observable<T>) => value is S, resultSelector: void, defaultValue?: S): Observable<S>;
   <R>(predicate?: (value: T, index: number, source: Observable<T>) => boolean, resultSelector?: (value: T, index: number) => R,
       defaultValue?: R): Observable<R>;
+  // We can cast `T -> R` in it if we pass the result selector.
```

Comment:
it's just nitpicking, but I'm feeling this'd be better to be interface level comment (@ line 61) - what do you think?